src/rebar_app_discover.erl

```diff
@@ -34,13 +34,15 @@ format_error({missing_module, Module}) ->
     io_lib:format("Module defined in app file missing: ~p~n", [Module]).
 
 merge_deps(AppInfo, State) ->
-    Profiles = rebar_state:current_profiles(State),
+    Default = rebar_state:default(State),
+    CurrentProfiles = rebar_state:current_profiles(State),
     Name = rebar_app_info:name(AppInfo),
     C = rebar_config:consult(rebar_app_info:dir(AppInfo)),
 
+    %% We reset the opts here to default so no profiles are applied multiple times
     AppState = rebar_state:apply_overrides(
                  rebar_state:apply_profiles(
-                   rebar_state:new(State, C, rebar_app_info:dir(AppInfo)), Profiles), Name),
+                   rebar_state:new(rebar_state:opts(State, Default), C, rebar_app_info:dir(AppInfo)), CurrentProfiles), Name),
```

Comment:
`add_to_profile` should only being run in a providers `init`, which is what the fixes in this PR take into account.

It may be ok for `add_to_profile` to modify default in addition to opts instead, not sure.